pkg/controller/nexus/update/updates.go

```diff
@@ -0,0 +1,156 @@
+//     Copyright 2020 Nexus Operator and/or its authors
+//
+//     This file is part of Nexus Operator.
+//
+//     Nexus Operator is free software: you can redistribute it and/or modify
+//     it under the terms of the GNU General Public License as published by
+//     the Free Software Foundation, either version 3 of the License, or
+//     (at your option) any later version.
+//
+//     Nexus Operator is distributed in the hope that it will be useful,
+//     but WITHOUT ANY WARRANTY; without even the implied warranty of
+//     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//     GNU General Public License for more details.
+//
+//     You should have received a copy of the GNU General Public License
+//     along with Nexus Operator.  If not, see <https://www.gnu.org/licenses/>.
+
+package update
+
+import (
+	ctx "context"
+	"fmt"
+	"github.com/heroku/docker-registry-client/registry"
+	"github.com/m88i/nexus-operator/pkg/apis/apps/v1alpha1"
+	"github.com/m88i/nexus-operator/pkg/logger"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	communityNexusRegistry     = "https://registry.hub.docker.com"
+	communityNexusRepo         = "sonatype/nexus3"
+	tagParseFailureFormat      = "unable to parse tag \"%s\": %v"
+	unableToCheckUpdatesFormat = "Unable to check for updates: %v"
+)
+
+var (
+	latestMicros = make(map[int]string)
+	// necessary to prevent this package from updating the tags while a caller tries to read from it
+	tagsMutex = sync.RWMutex{}
+
+	log = logger.GetLogger("updater")
+)
+
+// GetLatest returns the most recent image tag within a minor (the "y" in "x.y.z").
+// If the minor was not found, the second return value is false. It is expected callers check this.
+func GetLatest(minor int) (tag string, ok bool) {
+	tagsMutex.RLock()
+	defer tagsMutex.RUnlock()
+	tag, ok = latestMicros[minor]
+	return
+}
+
+// CheckUpdatesLoop checks if there is an update available every "delay" duration.
+// It also receives a client used to query the API server so it can start reconciliations for the controller when a newer tag for a given minor is found.
+func CheckUpdatesLoop(reconcileNexus reconcile.Reconciler, client client.Client, delay time.Duration) {
+	for {
+		time.Sleep(delay)
```

Comment:
I'm not sure if we should have a thread on going here since the reconcile loop will be called on t even if there's no change in the resources managed by the controllers. We could take an advantage from their timer instead of having ours.

We could continue having a go routine/new thread in the main reconciliation loop though, since this process should be async.

Reply:
Oh man, I had no idea that reconcile is called even if there is no change to the watched resources. This will simplify logic **a lot**. Here we only need to update the `latestMicros` map then. My idea is to call `GetLatest` from the Deployment Manager whenever creating the expected Deployment CR.

So this thread will run and from time to time check for updates and updating its internal map. Meanwhile, when a reconciliation begins, the Deployment Manager starts and defines the Deployment using the `GetLatest` function when it reaches the image definition. If it's different from the previous, it will update the Deployment and Kubernetes handles the rest.

Reply:
Oh, I think I misunderstood what you meant. I think you suggested calling the checker from the reconcile loop in order to fetch the tags only when a reconcile loop runs. Yeah, that's a good idea! Thanks for the suggestion